cpp/oneapi/dal/graph/test/service_functions_test.cpp

```diff
@@ -0,0 +1,187 @@
+      }
+      outf.close();
+    }
+
+    void delete_test_data() { std::remove(filename.c_str()); }
+
+    const std::string get_filename() const { return filename; }
+
+    std::int64_t get_correct_vertex_count() const { return vertex_count; }
+
+    std::int64_t get_correct_edge_count() const { return edge_count; }
+
+  protected:
+    std::string filename;
+    std::string file_content;
+    std::int64_t vertex_count;
+    std::int64_t edge_count;
+  };
+
+  class two_vertices_graph_type : public graph_base_data {
+  public:
+    two_vertices_graph_type() {
+      filename = "test_two_vertices_graph.csv";
+      file_content = "0 1";
+      vertex_count = 2;
+      edge_count = 1;
+    }
+
+    static const std::int64_t correct_vertex_count = 2;
+    static const std::int64_t neighbors_count = 2;
+    std::array<std::int64_t, correct_vertex_count> degrees = {1, 1};
+    std::array<std::int64_t, neighbors_count> neighbors = {1, 0};
+  };
+
+  class acyclic_graph_type : public graph_base_data {
+  public:
+    acyclic_graph_type() {
+      filename = "test_acyclic_graph.csv";
+      file_content = "0 1 0 2 0 4 2 3 2 6 3 5 3 7";
+      vertex_count = 8;
+      edge_count = 7;
+    }
+
+    static const std::int64_t correct_vertex_count = 8;
+    static const std::int64_t neighbors_count = 14;
+    std::array<std::int64_t, correct_vertex_count> degrees = {3, 1, 3, 3,
+                                                              1, 1, 1, 1};
+    std::array<std::int64_t, neighbors_count> neighbors = {1, 2, 4, 0, 0, 3, 6,
+                                                           2, 5, 7, 0, 3, 2, 3};
+  };
+
+  class complete_graph_type : public graph_base_data {
+  public:
+    complete_graph_type() {
+      filename = "test_complete_graph.csv";
+      file_content = "0 1 0 2 0 3 0 4 1 2 1 3 1 4 2 3 2 4 3 4";
+      vertex_count = 5;
+      edge_count = 10;
+    }
+
+    static const std::int64_t correct_vertex_count = 5;
+    static const std::int64_t neighbors_count = 20;
+    std::array<std::int64_t, correct_vertex_count> degrees = {4, 4, 4, 4, 4};
+    std::array<std::int64_t, neighbors_count> neighbors = {
+        1, 2, 3, 4, 0, 2, 3, 4, 0, 1, 3, 4, 0, 1, 2, 4, 0, 1, 2, 3};
+  };
+
+  class service_functions_test {
+  public:
+    using my_graph_type = dal::preview::undirected_adjacency_vector_graph<>;
+
+    template <typename GraphType> auto create_graph() {
+      GraphType graph_data;
+      graph_data.write_test_data();
+      const dal::preview::graph_csv_data_source ds(graph_data.get_filename());
+      const dal::preview::load_graph::descriptor<
+          dal::preview::edge_list<int32_t>, my_graph_type> desc;
+      auto my_graph = dal::preview::load_graph::load(desc, ds);
+      graph_data.delete_test_data();
+      return my_graph;
+    }
+
+    template <typename GraphType>
+    void check_vertex_degree(const GraphType &graph,
+                             std::int64_t *correct_degrees,

```

Comment:
If you use container, you can pass  it by const reference instead of casting to raw pointer.